multicast/src/main/kotlin/com/dropbox/flow/multicast/SharedFlowProducer.kt

```diff
@@ -42,34 +43,34 @@ internal class SharedFlowProducer<T>(
     private val src: Flow<T>,
     private val sendUpsteamMessage: suspend (ChannelManager.Message.Dispatch<T>) -> Unit
 ) {
-    private lateinit var collectionJob: Job
+    private val collectionJob: Job = scope.launch(start = CoroutineStart.LAZY) {
+        try {
+            src.catch {
+                sendUpsteamMessage(ChannelManager.Message.Dispatch.Error(it))
+            }.collect {
+                val ack = CompletableDeferred<Unit>()
+                sendUpsteamMessage(
+                    ChannelManager.Message.Dispatch.Value(
+                        it,
+                        ack
+                    )
+                )
+                // suspend until at least 1 receives the new value
+                ack.await()
+            }
+        } catch (closed: ClosedSendChannelException) {
+            // ignore. if consumers are gone, it might close itself.
+        }
+    }
 
     /**
      * Starts the collection of the upstream flow.
      */
     fun start() {
         scope.launch {
             try {
-                // launch again to track the collection job
-                collectionJob = scope.launch {
-                    try {
-                        src.catch {
-                            sendUpsteamMessage(ChannelManager.Message.Dispatch.Error(it))
-                        }.collect {
-                            val ack = CompletableDeferred<Unit>()
-                            sendUpsteamMessage(
-                                ChannelManager.Message.Dispatch.Value(
-                                    it,
-                                    ack
-                                )
-                            )
-                            // suspend until at least 1 receives the new value
-                            ack.await()
-                        }
-                    } catch (closed: ClosedSendChannelException) {
-                        // ignore. if consumers are gone, it might close itself.
-                    }
-                }
+                // start the collection
+                collectionJob.start()
```

Comment:
Calling `start` right before `join` is redundant. From [Job's KDoc](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-job/):
> Usually, a job is created in active state (it is created and started). However, coroutine builders that provide an optional start parameter create a coroutine in new state when this parameter is set to CoroutineStart.LAZY. Such a job can be made active by invoking start or join.